### ISSUE TITLE
slow startup typing performance when explorer contributions are expanded

### ISSUE THREAD

**Issue by chrisdias** [Microsoft/vscode#36575]
Windows, Insiders build 1.18.0-insiders commit 4b0145 date 2017-10-19

I noticed that on startup, there is a delay to typing.  The editor comes up, but when I start typing it takes a few seconds for the letters to appear.

Here's all the extensions I have installed

arcticicestudio.nord-visual-studio-code@0.6.0
benjpas.close-all@1.0.2
chrisdias.vscode-opennewinstance@0.0.4
christian-kohler.npm-intellisense@1.3.0
cssho.vscode-svgviewer@1.4.3
DotJoshJohnson.xml@1.9.2
eg2.vscode-npm-script@0.2.1
evilz.vscode-reveal@1.0.0
formulahendry.auto-close-tag@0.5.1
ms-azuretools.vscode-azureappservice@0.2.0
ms-azuretools.vscode-azurefunctions@0.1.0
ms-azuretools.vscode-cosmosdb@0.1.1
ms-vscode.azure-account@0.1.3
ms-vscode.azurecli@0.3.0
ms-vscode.PowerShell@1.4.3
PeterJausovec.vscode-docker@0.0.20

Profile with extensions:
[CPU-20171019T092459.zip](https://github.com/Microsoft/vscode/files/1399225/CPU-20171019T092459.zip)

[5c2a_main.cpuprofile.txt](https://github.com/Microsoft/vscode/files/1399203/5c2a_main.cpuprofile.txt)
[5c2a_renderer.cpuprofile.txt](https://github.com/Microsoft/vscode/files/1399204/5c2a_renderer.cpuprofile.txt)

- VSCode Version: <code>Code - Insiders 1.18.0-insider (4b01451cea60f32412098735709fddcac567b33d, 2017-10-19T08:29:07.096Z)</code>
- OS Version: <code>Windows_NT x64 10.0.15063</code>
- CPUs: <code>Intel(R) Core(TM) i5-2400S CPU @ 2.50GHz (4 x 2494)</code>
- Memory (System): <code>7.98GB (2.58GB free)</code>
- Memory (Process): <code>232.49MB working set (234.23MB peak, 200.54MB private, 40.71MB shared)</code>
- Load (avg): <code>0, 0, 0</code>
- VM: <code>0%</code>
- Initial Startup: <code>yes</code>
- Screen Reader: <code>no</code>
- Empty Workspace: <code>no</code>
- Timings:

|Component|Task|Time (ms)|
|---|---|---|
|main|start => app.isReady|37|
|main|app.isReady => window.loadUrl()|292|
|renderer|window.loadUrl() => begin to require(workbench.main.js)|442|
|renderer|require(workbench.main.js)|731|
|renderer|create extension host => extensions onReady()|612|
|renderer|restore viewlet|492|
|renderer|restore editor view state|1386|
|renderer|overall workbench load|1560|
|main + renderer|start => extensions ready|2271|
|main + renderer|start => workbench ready|3178|

Here is a profile without extensions loaded, where the performance is ok. So, it has to be an extension.

[Profile-no-extensions.zip](https://github.com/Microsoft/vscode/files/1399261/Profile-no-extensions.zip)

**Comment by chrisdias** [comment]
Ok, it's when there is an extension that has an explorer contribution, and the explorer contribution is expanded on startup.

To reproduce:

- Make sure you have docker installed, and have a couple of images on the machine (e.g. just do a `docker pull node:latest`)
- Install the [Docker extension](https://marketplace.visualstudio.com/items?itemName=PeterJausovec.vscode-docker)
- Open VS code, expand the DOCKER sash so that you see the "Images", "Containers", and "Registries" nodes.
- Open a file
- Close VS Code
- Restart VS Code
- As soon as you have a cursor, start typing

Result: huge delay until the Docker Explorer is initialized/visible.
Expected: I don't expect to be delayed by an extension loading.

If you don't want to install Docker, you can install the Azure CosmosDB Extension instead and try the same steps. The key is to have the explorer expanded so that the next time you start VS code it is expanded.

**Comment by alexandrudima** [comment]
Typing in the editor executes the `type` command through the command service (for the case of outsourcing typing to the extension host, e.g. the vim plugin).

The command service waits for the extension host to activate `*` extensions before invoking our built-in type command.

If there are many *slow* extensions, the extension host will take a while to be ready.

**Comment by jrieken** [comment]
> To double check if there is any correlation with the explorer tree besides that having that makes the extension host busy.

/cc @sandy081 for that

**Comment by sandy081** [comment]
Custom views in explorer are lazily activated - When the user expands the view or if it was already expanded before. Initial state is always collapsed.

After activating the extension contributing the view, it will ask the extension for the first level elements and updates the view.. it will also react to the requests coming from the extension (like refresh). Otherwise, explorer view should not be making the main/renderer process busy.

Have to investigate to get more insights.

**Comment by sandy081** [comment]
Some details I observed:

I see some key-strokes lagging with expanded docker view when compared with that of collapsed. For eg: In former, I see text after ~5-6 key strokes and in latter it is after ~2-3. The delta time could be spent in rendering the first level elements of the tree.

I do not see this delta with other custom view extensions like `Gitlens` and `Github Issues and PRs`. One difference I see between these and `Docker` extension is that, docker extension has static first level data and others fetch using git or github.

I am assuming that, since Docker is returning static first level elements, Tree rendering could be happening immediately after extension host is ready and hence the key strokes are lagging little.

Not sure if we can delay rendering of the view further.

**Comment by alexandrudima** [comment]
@jrieken Maybe we can do the following change to the command service:

When a request comes in to run a command:
* if the extension host is not ready yet and if the command is defined in the commands map
* we kick off an activation event for that command, but don't wait for it and execute the defined command.

Otherwise, I'm going to have to take the editor's `type` command out of using the command service, but then I'm not sure what else to use such as to not break the vim case.

**Comment by jrieken** [comment]
@alexandrudima didn't you already make that change? https://github.com/Microsoft/vscode/blob/master/src/vs/platform/commands/common/commandService.ts#L34

**Comment by alexandrudima** [comment]
@jrieken You're right... 

Perhaps what's happening here is that the extension host is ready, but running the activation event `onCommand:type` takes a long time on the extension host, presumably ... because the extension host is super busy doing tree/explorer stuff ?

**Comment by jrieken** [comment]
Yeah, that's good guess. I have seen the extension host starting quickly but activation taking much longer often...

**Comment by alexandrudima** [comment]
I'll look into avoiding going to the extension host for activation events which no extension has registered for. This should alleviate the problem with `onCommand:type`.

**Comment by alexandrudima** [comment]
Ok this is bizzare. That optimization is already implemented. Activation events don't go to the extension host if there are no extensions registered for them

https://github.com/Microsoft/vscode/blob/fa4c5561c9485a3c4427ec3de5584b9334e3e3c4/src/vs/workbench/services/extensions/electron-browser/extensionService.ts#L222

**Comment by alexandrudima** [comment]
Ok, not so bizzare. emmet listens to `"onCommand:type"` which forces that this activation event gets sent over to the extension host. So, effectively, since emmet is a built-in extension, you cannot type in the editor until emmet activates.

**Comment by alexandrudima** [comment]
I have created PR #36876 for @ramya-rao-a to look into.

I consider my side of things working ok.

**Comment by sandy081** [comment]
@alexandrudima Dies this PR fixes this issue or issue still remains?

**Comment by alexandrudima** [comment]
@sandy081 The PR makes it that typing in the editor is not blocked by the extension host out-of-the-box. The PR does not change anything else. I leave it up to you if you want to investigate why expanding a tree in the explorer would make the extension host a noticeable number of seconds slower.

**Comment by sandy081** [comment]
Closing this issue as the PR from @alexandrudima fixes the reported issue. Filed a new issue https://github.com/Microsoft/vscode/issues/37393 for investigating the point raised in above comment.

**Comment by chrmarti** [comment]
Verified in code and couldn't reproduce any delay.
